pyomo/contrib/piecewise/triangulations.py

```diff
@@ -0,0 +1,723 @@
+
+
+# get the v_0 and sign vectors corresponding to a given square, identified by its
+# low-coordinate corner
+def _get_nearest_odd_and_sign_vec(corner):
+    v_0 = []
+    sign = []
+    for x in corner:
+        if x % 2 == 0:
+            v_0.append(x + 1)
+            sign.append(-1)
+        else:
+            v_0.append(x)
+            sign.append(1)
+    return v_0, sign
+
+
+def _get_grid_hamiltonian(dim, length):
+    if dim == 1:
+        return [[n] for n in range(length)]
+    else:
+        ret = []
+        prev = _get_grid_hamiltonian(dim - 1, length)
+        for n in range(length):
+            # if n is even, add the previous hamiltonian with n in its new first
+            # coordinate. If odd, do the same with the previous hamiltonian in reverse.
+            if n % 2 == 0:
+                for x in prev:
+                    ret.append([n] + x)
+            else:
+                for x in reversed(prev):
+                    ret.append([n] + x)
+        return ret
+
+
+# Fix vertices (in place) when the simplices are right but vertices are not
+def _fix_vertices_incremental_order(simplices):
+    last_vertex_index = len(simplices[0]) - 1
+    for i, simplex in enumerate(simplices):
+        # Choose vertices like this: first is always the same as last
+        # of the previous simplex. Last is arbitrarily chosen from the
+        # intersection with the next simplex.
+        first = None
+        last = None
+        if i == 0:
+            first = 0
+        else:
+            first = simplex.index(simplices[i - 1][last_vertex_index])
+
+        if i == len(simplices) - 1:
+            last = last_vertex_index
+        else:
+            for n in range(last_vertex_index + 1):
+                if simplex[n] in simplices[i + 1] and n != first:
+                    last = n
+                    break
+        if first is None or last is None:
+            raise DeveloperError("Couldn't fix vertex ordering for incremental.")
+
+        # reorder the simplex with the desired first and last
+        new_simplex = list(simplex)
+        temp = new_simplex[0]
+        new_simplex[0] = new_simplex[first]
+        new_simplex[first] = temp
+        if last == 0:
+            last = first
+        temp = new_simplex[last_vertex_index]
+        new_simplex[last_vertex_index] = new_simplex[last]
+        new_simplex[last] = temp
+        simplices[i] = tuple(new_simplex)
+
+
+# Let G_n be the graph on n! vertices where the vertices are permutations in
+# S_n and  two vertices are adjacent if they are related by swapping the values
+# of pi(i - 1) and pi(i) for some i in {2, ..., n}.
+#
+# This function gets a Hamiltonian path through G_n, starting from a fixed
+# starting permutation, such that a fixed target symbol is either the image
+# rho(1), or it is rho(n), depending on whether first or last is requested,
+# where rho is the final permutation.
+def _get_Gn_hamiltonian(n, start_permutation, target_symbol, last, _cache={}):
```

Comment:
This is dangerous (although maybe you intend it?): all calls to `_get_Gn_hamiltonian` that do not provide `_cache` share the same dict object.

Reply:
I'm pretty sure this is intended. But maybe @sadavis1 can confirm at some point... (And perhaps we should add a comment).